### ISSUE TITLE
pl.Categorical columns can't be converted back to arrow.

### ISSUE THREAD

**Issue by ghuls** [pola-rs/polars#672]
#### Are you using Python or Rust?

**Python**.

#### Which feature gates did you use?

This can be ignored by Python users.

#### What version of polars are you using?

0.7.16

#### What operating system are you using polars on?

Ubuntu 20.04

#### Describe your bug.

`pl.Categorical` columns can't be converted back to arrow.

`RuntimeError: Any(ArrowError(InvalidArgumentError("column types must match schema types, expected UInt16 but found UInt32 at column index 0")))`

Arrow has dictionary encoding for representing categorical data I think.
https://github.com/apache/arrow/search?q=categorical

#### What are the steps to reproduce the behavior?

```python
In [525]: pl.DataFrame({
     ...:     'col1' : np.array(['a', 'b', 'c', 'd', 'e']),
     ...:     'col2': np.array([0, 1, 0, 1, 0])
     ...: }).to_arrow()
Out[525]: 
pyarrow.Table
col1: large_string
col2: int64

In [526]: pl.DataFrame({
     ...:     'col1' : np.array(['a', 'b', 'c', 'd', 'e']),
     ...:     'col2': np.array([0, 1, 0, 1, 0])
     ...: }).with_column(pl.col('col1').cast(pl.Categorical))
Out[526]: 
shape: (5, 2)
╭──────┬──────╮
│ col1 ┆ col2 │
│ ---  ┆ ---  │
│ cat  ┆ i64  │
╞══════╪══════╡
│ "a"  ┆ 0    │
├╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ "b"  ┆ 1    │
├╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ "c"  ┆ 0    │
├╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ "d"  ┆ 1    │
├╌╌╌╌╌╌┼╌╌╌╌╌╌┤
│ "e"  ┆ 0    │
╰──────┴──────╯

In [527]: pl.DataFrame({
     ...:     'col1' : np.array(['a', 'b', 'c', 'd', 'e']),
     ...:     'col2': np.array([0, 1, 0, 1, 0])
     ...: }).with_column(pl.col('col1').cast(pl.Categorical)).to_arrow()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-527-d121d3997771> in <module>
----> 1 pl.DataFrame({
      2     'col1' : np.array(['a', 'b', 'c', 'd', 'e']),
      3     'col2': np.array([0, 1, 0, 1, 0])
      4 }).with_column(pl.col('col1').cast(pl.Categorical)).to_arrow()

~/software/anaconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/frame.py in to_arrow(self)
    279         This operation is zero copy.
    280         """
--> 281         record_batches = self._df.to_arrow()
    282         return pa.Table.from_batches(record_batches)
    283 

RuntimeError: Any(ArrowError(InvalidArgumentError("column types must match schema types, expected UInt16 but found UInt32 at column index 0")))
```

**Comment by ghuls** [comment]
```python
In [533]: ?c.dictionary_encode
Docstring:
ChunkedArray.dictionary_encode(self, null_encoding=u'mask')

Compute dictionary-encoded representation of array

Returns
-------
pyarrow.ChunkedArray
    Same chunking as the input, all chunks share a common dictionary.
Type:      builtin_function_or_method

In [534]: b = pl.DataFrame({
     ...:     'col1' : np.array(['a', 'b', 'c', 'd', 'e', 'd', 'd', 'b', 'c',]),
     ...:     'col2': np.array([0, 1, 0, 1, 0, 2 ,3 ,4, 5])
     ...: }).to_arrow()

In [535]: b
Out[535]: 
pyarrow.Table
col1: large_string
col2: int64

In [536]: c = b.column(0)

In [537]: c.dictionary_encode()
Out[537]: 
<pyarrow.lib.ChunkedArray object at 0x7f52a305a180>
[

  -- dictionary:
    [
      "a",
      "b",
      "c",
      "d",
      "e"
    ]
  -- indices:
    [
      0,
      1,
      2,
      3,
      4,
      3,
      3,
      1,
      2
    ]
]
```

**Comment by ritchie46** [comment]
The Error should be fixed by #672. However there is quite a big gotcha with `Categorical` date and `arrow`. Polars does not use the `dictionary-encoded` array. I chose not to because reverse lookups (from key to string) are `O(n)` in these arrays. 

For this reason I used by own representation for dictionary encoded data. This makes it more useful with a large number of unique `string` values. However, it does mean that if we want to go to `py-arrow` and back we need to `coerce` from `Categorical` to `Utf8` before we sent to `py-arrow` and back to `Categorical` when sending back.

**Comment by ghuls** [comment]
> For this reason I used by own representation for dictionary encoded data. This makes it more useful with a large number of unique `string` values. However, it does mean that if we want to go to `py-arrow` and back we need to `coerce` from `Categorical` to `Utf8` before we sent to `py-arrow` and back to `Categorical` when sending back.

Seems like the coercing is not done at the moment even tough the help seems to indicate it.

```python
In [23]: ?region_ids_crm_scores_and_rankings.to_arrow
Signature: region_ids_crm_scores_and_rankings.to_arrow() -> pyarrow.lib.Table
Docstring:
Collect the underlying arrow arrays in an Arrow Table.
This operation is mostly zero copy.

Data types that do copy:
    - CategoricalType
File:      ~/software/anaconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/frame.py
Type:      method

In [25]: region_ids_crm_scores_and_rankings.dtypes
Out[25]: 
[polars.datatypes.Categorical,
 polars.datatypes.Float32,
 polars.datatypes.UInt32,
 polars.datatypes.UInt32]

In [26]: region_ids_crm_scores_and_rankings.to_arrow()
Out[26]: 
pyarrow.Table
region_ids: uint32
score: float
ranking: uint32
ranking_to_high: uint32
```

I think it would be useful to have an option to coerce pl.Catergorical to arrow dictionary encoded and just string representation.
